Rules to support running setup.py commands

benjyw opened the pull request.

benjyw commented:
Reviewers: This is a doozy, but the individual commits are structured to be independently reviewable, in order, and I recommend doing so!

Please comment on any test cases you think I should add (looking at the old tests/python/pants_test/backend/python/tasks/test_setup_py.py and tests/python/pants_test/backend/python/tasks/test_setup_py_integration.py may be instructive). However, unlike those tests, these ones are a lot more modular, thanks to the modularity of the underlying rules, so many more edge cases are already covered.

I have also tested this manually.

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py_test.py:
```diff
@@ -54,3 +54,150 @@ def alias_groups(cls):
+      OwnedDependencies, Params(DependencyOwner(ExportedTarget(self.tgt(exported))))
+    ))
+
+  def test_owned_dependencies(self):
```
Nit: adding `-> None` to all these tests will ensure MyPy checks the implementation of the functions.

benjyw replied:
Done. Why wouldn't MyPy otherwise check them?

Eric-Arellano replied:
Incremental type hinting. MyPy doesn't check function bodies unless you tell it to by adding at least one type hint.

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py_test.py:
```diff
@@ -56,6 +56,64 @@ def tgt(self, addr: str) -> HydratedTarget:
+      RootRule(SourceRootConfig),
+    ]
+
+  def assert_ancestor_init_py(self, expected_init_pys, addrs):
```
Type hints appreciated. Think this would be

```python
def assert_ancestor_init_py(self, expected_init_pys: Iterable[str], addrs: Iterable[str]) -> None:
```

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
+    register(
+      '--args', type=list, member_type=str,
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
```
Even better would be saying `--setup-py2-args='bdist_wheel --python-tag py36.py37'`. The reason I recommend this is to clarify that the string gets shlexed.

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
+      setup_py_result = await Get[RunSetupPyResult](
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
```
This won't work properly with shlexing. Instead, define on the `GoalSubsystem` a function `get_args()` that does this:

https://github.com/pantsbuild/pants/blob/a6716b441e260a5c8cce663d4c4df76c4c4d7a14/src/python/pants/backend/python/lint/black/subsystem.py#L31-L32

Fwit, the refactoring I've been doing on `option/` the past few days has been prework to move support for shlexable strings directly into the parser itself to avoid this "gotcha".

Eric-Arellano replied:
Once we merge https://github.com/pantsbuild/pants/pull/8904, this will be irrelevant. I'm planning on merging once I get one more review, so probably not worth your time to add `get_args()` and then have to delete it tomorrow.

benjyw replied:
Done! And I stand chastised by my falling for the gotcha I termed "minor"...

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
+    workspace.materialize_directory(
+      DirectoryToMaterialize(result_digest, path_prefix=str(setup_py_dir))
+    )
+    console.print_stderr(f'Wrote setup.py chroot for {addr} to {setup_py_dir}')
```
Why stderr?

stuhood replied:
+1. This feels like the actual result of the rule, and so should be stdout. And unlike the above, it _should_ probably go to the console, because it will/should happen every time the console_rule runs.

benjyw replied:
I generally think of stdout as for the formal output of a thing, and stderr for logging/information. So a rule whose stated purpose is to output some data (e.g., `goals`) should do that to stdout. But this rule's purpose is not to output anything, this output is just incidental. For example, we could change its phrasing without incurring a deprecation cycle.

But probably it should just be logging?

benjyw replied:
Naa, I tried logging and it doesn't look right. I think directly writing to stderr is the right thing here.

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+    # TODO: Implement Python 2 support.  This will involve, among other things: merging ancestor
+    # __init__.py files into the chroot, detecting packages based on the presence of __init__.py,
+    # and inspecting all __init__.py files for the namespace package incantation.
+    raise UnsupportedPythonVersion('Running setup.py commands not supported for Python 2.')
```
To clarify, will Python 3 work when not using PEP 420? It's valid to have a Python 3-only codebase that doesn't use implicit namespace packages, e.g. the Pants repo.

Either way, totally okay with punting on this functionality. Only want to clarify what the missing functionality is and possibly update the error message.

benjyw replied:
Yes, that should work. What won't work is python3 but with explicit namespace package incantations in `__init__.py` files. I'm not sure those are allowed in python3 anyway though. PEP420 implies that they are not.

benjyw replied:
In other words, I think you are always "using PEP 420" in python3. The PEP allows for non-namespace packages, by having explicit `__init__.py` files. What it doesn't appear to allow is for those `__init__.py` files to create namespace packages using the python2-style.

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
```
The `await Get`s inside of your loop inhibit parallelism (as each target will be processed sequentially). To allow for parallelism, you'd want to move toward using `MultiGet` for more of this. See #7491 for more info.

benjyw replied:
Yeah, I went back and forth on this, mostly for UX reasons. But now I recall based on a comment in the original setup.py task that at least the setup.py executions themselves must run sequentially, in dep order, because if exportable target B depends on exportable target A, and the setup.py command we run involves publishing (e.g., `sdist upload`), then we don't want to publish B before A.

Maybe the above behavior should be controlled by an option, so that we can get concurrency when the publishing dependency thing isn't an issue. I'll address this one way or another.

stuhood replied:
> and the setup.py command we run involves publishing (e.g., sdist upload), then we don't want to publish B before A.

...oof. That's scary.

A strategy that would still work for that case without being embarassingly parallel would be to do this recursively (setup.py for X depends on setup.py for Y).

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
```
This might be a lie due to caching: would recommend moving it to a log statement inside of the rule that actually prepares the execution... which might _still_ be a lie due to the persistent cache, but which would at least avoid rendering it for every invoke under pantsd.

benjyw replied:
I'll drop it entirely, for now anyway. The v2 ui will tell us when a process is actually running.

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+    pex_path="./setuptools.pex",
+    pex_args=('setup.py', *req.args),
+    input_files=merged_input_files,
+    output_files=('**',),
```
Hm... globs are not supported by the remote execution API. This works because we use our glob expansion code to capture outputs locally, but we should probably make globs an error in `ExecuteProcessRequest`.

But also: you probably don't want to actually capture _everything_ here, do you? You'll capture the pex you were running, for example. Consider writing to an output directory and then capturing that?

benjyw replied:
Good to know. So if I don't know the names of output files in advance I use `output_directories`? 

You're right that I don't want to capture everything. But I'm not sure what the correct surface area is.  All dists created by running`setup.py` commands should be under `dist/` in the chroot, so maybe that's enough for now.

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+  # Find the ancestors of all dirs containing .py files, including those dirs themselves.
+  source_dir_ancestors: Set[Tuple[str, str]] = set()  # Items are (src_root, path incl. src_root).
+  sources_snapshots = await MultiGet(
+    Get[Snapshot](Digest, target.adaptor.sources.snapshot.directory_digest)
```
`target.adaptor.sources.snapshot` is already a `Snapshot`, so you shouldn't need to re-load it here.

benjyw replied:
🤦‍♂

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+  return getattr(target.adaptor, 'provides', None) is not None
+
+
+@rule(name="Get requirements")
```
This isn't a particularly meaningful name... is this something like "Compute transitive 3rdparty requirements"? Also, @gshuflin : is there a convention yet for which part-of-speech/tense a name should be? ie "Compute" vs "Computing"

benjyw replied:
I went with "Compute distribution's 3rd party requirements"

Will wait re part-of-speech

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+
+  # Get all 3rdparty deps of our owned deps.
+  #
+  # Note that we need only consider requirements that are direct dependencies of our owned deps:
```
Hm. "strict deps" are not generally a requirement within Pants python codebases... is this a change in behaviour, or is this how v1 worked in this case?

benjyw replied:
This doesn't actually require strict deps, and is how v1 worked:  

Say exported target C depends on exported target A and regular target B.  In other words, C's dist publishes the code from C and B, and A's dist publishes the code from A. 

So we do want C and B's 3rdparty requirements to become the requirements of C's dist, but we do *not* want A's requirements to (which is why we don't just take all transitive deps). So if C references a 3rdparty symbol that it depends on only via A, distC won't list that requirement (but distA will, and distC will list distA, so the non-strict effective dep will be preserved). If you want your dists to have strict deps, you need your targets to as well, at least across exported target boundaries.

stuhood commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,532 @@
+  return ExportedTargetRequirements(tuple(sorted(req_strs)))
+
+
+@rule(name="Get owned targets")
```
Ditto above. Also, it's worth considering in cases like this where the logic is abstract whether the rule is even worth naming.

benjyw replied:
I figured it's worth naming logic that might be time-consuming, so users can see what's going on.  Since these operations might take time on a large graph, I figured it was worth naming them. But we can revisit.

benjyw commented on src/python/pants/backend/awslambda/common/awslambda_common_rules.py:
```diff
@@ -31,7 +31,6 @@ class AWSLambdaTarget:
 
 
 class AWSLambdaOptions(LineOriented, GoalSubsystem):
-  """Runs tests."""
```
The description comes from the GoalSubsystem docstring, not the Goal docstring?

Eric-Arellano replied:
Yes, to mirror how normal Subsystems work. The idea of GoalSubsystem is to the be the public API for your goal, meaning it has the help message and any options. 

Goal is only a simple dataclass with one field of exit_code that gets returned by your console_rule.

stuhood replied:
(the split is more confusing than the previous factoring, imo)

benjyw commented on src/python/pants/backend/python/rules/setuptools.py:
```diff
@@ -0,0 +1,12 @@
+
+class Setuptools(PythonToolBase):
+  # NB: setuptools doesn't have an entrypoint, unlike most python tools.
+  # We call it via a generated setup.py script.
```
As discussed in that issue, a better solution is to refactor mixins and base classes appropriately. In this case we could either create two flavors of PythonToolBase, or just allow entrypoint to be None (which is the case today de-facto, so we'd just need to document it).

benjyw replied:
In fact I will add the relevant documentation now.

benjyw commented on src/python/pants/backend/python/rules/setuptools.py:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
```
Not sure I understand what else would go in a `backend/python/setuptools` directory?  I'm also not sure that having `rules` and `subsystems` subdirs is useful in general, although we were ~rigid about this with v1 (`subsystems`, `tasks`, `targets` subdirs everywhere). 

Why would `register.py` need to import this? In v2 we don't need to register subsystems, afaik? The engine does this?

benjyw commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -48,3 +48,26 @@
+      interpreter_constraints=PexInterpreterConstraints(
+        constraint_set=tuple(setuptools.default_interpreter_constraints)
+      ),
+      entry_point=setuptools.get_entry_point(),
```
You're right, that was copy-pasta. Good catch!

benjyw commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -49,12 +49,121 @@
+  """
+  tht = await Get[TransitiveHydratedTargets](
+    BuildFileAddresses([dependency_owner.exported_target.hydrated_target.address]))
+  all_tgts = list(tht.closure)
```
I think that was left over from a previous iteration, when I needed a stable order. Now I actually don't. 

But generally, I don't know how one can be sure that `TransitiveHydatedTargets.closure` is an OrderedSet. The type annotation is `Any`.  So nothing formal appears to guarantee stable order.

Eric-Arellano replied:
It’s only Any because we haven’t yet added proper annotations. Also, we can’t use typed OrderedSets yet. We can declare it’s an OrderedSet but not what the element type is. 

Fwit, in my free time I put up a PR to the ordered-set library to improve distribution of their type stubs. If that lands, I’ll put up a PR to replace Twitter Commons version with this library.

benjyw commented:
OK, I've addressed all comments. I've made the chroot creating concurrent using MultiGet, but not the actual running of setup.py. I've left a TODO about what needs to happen there. Will ponder this and address today.

Also added a line to write a MANIFEST.in, to ensure that setup.py itself is included in sdists.

PTAL, thanks!

Eric-Arellano commented on src/python/pants/backend/python/rules/run_setup_py_test.py:
```diff
@@ -0,0 +1,482 @@
+      RootRule(SourceRootConfig)
+    ]
+
+  def assert_chroot(self, expected_files, expected_setup_kwargs, addr):
```
Style difference: I find kwargs much easier to understand because then I don't need to keep referencing this function definition to know what arguments you're passing to `assert_chroot`. I recommend enforcing kwargs via:

```python
  def assert_chroot(self, *, expected_files: List[str], expected_setup_kwargs: Dict, addr: str) -> None:
```

benjyw commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,552 @@
+  for target, chroot in zip(exported_targets, chroots):
+    addr = target.hydrated_target.address.reference()
+    provides = target.hydrated_target.adaptor.provides
+    args = options.get_args()
```
> Now just `options.args`

options.values.args I think

benjyw commented:
> Yes, they are allowed in PY3 afaict. For example, Pants uses >__import__("pkg_resources").declare_namespace(__name__) and does not use PEP 420.

This may work in practice, but PEP 420 states, in https://www.python.org/dev/peps/pep-0420/#specification:

```
Regular packages will continue to have an __init__.py and will reside in a single directory.

Namespace packages cannot contain an __init__.py. As a consequence, pkgutil.extend_path and pkg_resources.declare_namespace become obsolete for purposes of namespace package creation. There will be no marker file or directory for specifying a namespace package.
```

This is somewhat contradicted by https://www.python.org/dev/peps/pep-0420/#migrating-from-legacy-namespace-packages:

```
Because it is likely not practical for all existing portions of a namespace package to be migrated to this PEP at once, extend_path() will be modified to also recognize PEP 420 namespace packages.
```

But in any case this says nothing about `declare_namespace` in setuptools, which is what we use in lieu of extend_path.  

Basically, it works in practice but is *not* the supported way of doing namespace packages in python 3, if PEP 420 means anything.  Possibly we should still support it, but that will have to wait for a followup. For now I'll clarify the TODO.

benjyw commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -0,0 +1,552 @@
+  sources = await Get[SetupPySources](SetupPySourcesRequest(targets))
+  requirements = await Get[ExportedTargetRequirements](DependencyOwner(request.exported_target))
+
+  # Nest the sources under the src/ prefix.
```
Just for good order, AFAIK.

benjyw commented:
Issue to track TODOs here: https://github.com/pantsbuild/pants/issues/8912

jsirois commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -152,6 +163,47 @@ def _is_exported(target: HydratedTarget) -> bool:
+  # Add the requirements on any exported targets on which we depend.
+  exported_targets_we_depend_on = await MultiGet(
+    Get[ExportedTarget](OwnedDependency(ht)) for ht in owned_by_others)
+  req_strs.extend(et.hydrated_target.adaptor.provides.key
```
Not yours, but `key` is an odd name for this `PythonArtifact` property. It should be named something like `requirement` since it is , in fact, a pinned requirement - which is what we want. In setuptools parlance `Requirement` has a `key` and that is the bare project name. That would not be what we want here since it floats!

benjyw replied:
Yeah, that is weird and tripped me up slightly. I'll change in a followup.

benjyw replied:
But I'll change it in a followup I think, since it also affects v1.

jsirois commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -128,12 +136,94 @@ class SetupPySources:
+    'packages': sources.packages,
+    'namespace_packages': sources.namespace_packages,
+    'package_data': dict(sources.package_data),
+    'install_requires': sorted(requirements.requirement_strs)
```
This sorting does not respect declration order which is sometimes significant and is in contrast to v1. The surprise this creates would be getting a different resolve for `./pants repl <exported dist target>` for example from `pip install <exported dist>`. The `requirements.requirement_strs` should be stable so I think the sorting can just be eliminated.

benjyw replied:
Good catch. In fact, I was also sorting on creation of the ExportedTargetRequirements instance, which is also wrong. Removed that as well. Now the order will be whatever ` PexRequirements.create_from_adaptors()` returns, with the sorted exported target reqs appended. I assume `PexRequirements.create_from_adaptors()` order is stable?

jsirois commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
+      result_digest = setup_py_result.output
+    else:
+      # Just dump the chroot.
```
I like this dump vs args instead of v1 sdist vs args.

jsirois commented on src/python/pants/backend/python/rules/run_setup_py.py:
```diff
@@ -266,7 +277,9 @@ class SetupPy(Goal):
-    output_files=('**',),
+    # setuptools commands that create dists write them to dist/.
+    # TODO: Could there be other useful files to capture?
+    output_directories=('dist/',),
```
This is only the default. The `sdist` and `bdist_wheel` commands both accept a `--dist-dir` argument. Probably worth a TODO + issue to follow up and address later.

benjyw replied:
Added this to https://github.com/pantsbuild/pants/issues/8912

illicitonion commented:
I haven't reviewed, but this is very exciting! Is this sufficient that we can replace the v1 setup-py calls in our own release scripts with the v2 versions? Because that would make me very, very happy, and I suspect may shave 1-3 minutes from release candidate generation :)

benjyw commented on src/python/pants/backend/python/rules/setup_py_util.py:
```diff
@@ -0,0 +1,139 @@
+        # don't want to list the root package in the metadata.
+        packages.add(package)
+
+  # See which packages are namespace packages.
```
There is a TODO to that effect below, added a link.

benjyw commented:
> I haven't reviewed, but this is very exciting! Is this sufficient that we can replace the v1 setup-py calls in our own release scripts with the v2 versions? Because that would make me very, very happy, and I suspect may shave 1-3 minutes from release candidate generation :)

We'd still need to address the second issue in https://github.com/pantsbuild/pants/issues/8912 before we can build our dists (at least, with correct `namespace_packages` metadata, I'm not sure how important that is).

benjyw commented:
@stuhood This OK to ship?
